src/client_side.cc

```diff
@@ -3150,15 +3172,25 @@ ConnStateData::startPeekAndSplice()
     bio->setReadBufData(inBuf);
     bio->hold(true);
 
-    // Here squid should have all of the client hello message so the
-    // tlsAttemptHandshake() should return 0.
-    // This block exist only to force openSSL parse client hello and detect
-    // ERR_SECURE_ACCEPT_FAIL error, which should be checked and splice if required.
-    if (tlsAttemptHandshake(this, nullptr) < 0) {
-        debugs(83, 2, "TLS handshake failed.");
+    // We have successfully parsed client Hello, but our TLS handshake parser is
+    // forgiving. Now we use a TLS library to parse the same bytes, so that we
+    // can honor on_unsupported_protocol if needed.
+    //
+    // We should have just the client Hello so the tlsAttemptHandshake() should
+    // ask for more data (XXX: why?). And see TODO above the
+    // tlsAttemptHandshake() function definition.
+    const auto handshakeResult = tlsAttemptHandshake(this, nullptr);
+    if (handshakeResult.category == Security::IoResult::ioError) {
+        updateError(ERR_SECURE_ACCEPT_FAIL, handshakeResult.errorDetail);
```

Comment:
The old tlsAttemptHandshake will return:
 * 1 if the initial tls negotiation was finished. This is requires the client send its hello and squid respond with its server hello.
 * 0 if squid server needs more data, or squid server wants to write data (eg server hello repsonse)
 * -1 on error

Because we read all of the client hello message, but we did not send anything back yet, we are expecting that the (old) tlsAttemptHandshake, if no error, will return 0 because wants to write server hello to the client.

The comment:
>     // We should have just the client Hello so the tlsAttemptHandshake() should
>     // ask for more data (XXX: why?).

Should re-written to something lile:
>     // We should have just the client Hello so the tlsAttemptHandshake() should
>     // ask to write server hello response.

Reply:
Thank you for clarifying this. Polished in branch commit b1033eb.